daemon/api_themes.go

```diff
@@ -0,0 +1,245 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package daemon
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"path/filepath"
+	"regexp"
+	"sort"
+	"strings"
+
+	"github.com/snapcore/snapd/i18n"
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/overlord/auth"
+	"github.com/snapcore/snapd/overlord/snapstate"
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/store"
+	"github.com/snapcore/snapd/strutil"
+)
+
+var (
+	themesCmd = &Command{
+		Path:   "/v2/themes",
+		UserOK: true,
+		GET:    checkThemes,
+		POST:   installThemes,
+	}
+)
+
+type themeStatus string
+
+const (
+	themeInstalled   themeStatus = "installed"
+	themeAvailable   themeStatus = "available"
+	themeUnavailable themeStatus = "unavailable"
+)
+
+type themeStatusResponse struct {
+	GtkThemes   map[string]themeStatus `json:"gtk-themes"`
+	IconThemes  map[string]themeStatus `json:"icon-themes"`
+	SoundThemes map[string]themeStatus `json:"sound-themes"`
+}
+
+func getInstalledThemes(d *Daemon) (gtkThemes, iconThemes, soundThemes []string, err error) {
+	infos := d.overlord.InterfaceManager().Repository().Info(&interfaces.InfoOptions{
+		Names: []string{"content"},
+		Slots: true,
+	})
+	for _, info := range infos {
+		for _, slot := range info.Slots {
+			var content string
+			// The content interface ensures this attribute exists
+			if err := slot.Attr("content", &content); err != nil {
+				return nil, nil, nil, err
+			}
+			var themes *[]string
+			switch content {
+			case "gtk-3-themes":
+				themes = &gtkThemes
+			case "icon-themes":
+				themes = &iconThemes
+			case "sound-themes":
+				themes = &soundThemes
+			default:
+				continue
+			}
+			var sources []interface{}
+			if err := slot.Attr("source.read", &sources); err != nil {
+				continue
+			}
+			for _, s := range sources {
+				if path, ok := s.(string); ok {
+					*themes = append(*themes, filepath.Base(path))
+				}
+			}
+		}
+	}
+	sort.Strings(gtkThemes)
+	sort.Strings(iconThemes)
+	sort.Strings(soundThemes)
+	return gtkThemes, iconThemes, soundThemes, nil
+}
+
+var badPkgCharRegexp = regexp.MustCompile(`[^a-z]+`)
+
+func themePackageCandidates(prefix, themeName string) []string {
+	themeName = strings.ToLower(themeName)
+	themeName = badPkgCharRegexp.ReplaceAllString(themeName, "-")
+	themeName = strings.Trim(themeName, "-")
+
+	var packages []string
+	for themeName != "" {
+		packages = append(packages, prefix+themeName)
+		pos := strings.LastIndexByte(themeName, '-')
+		if pos < 0 {
+			break
+		}
+		themeName = themeName[:pos]
+	}
+	return packages
+}
+
+func getThemeStatusForType(ctx context.Context, theStore snapstate.StoreService, user *auth.UserState, prefix string, themes, installed []string, toInstall map[string]bool) (map[string]themeStatus, error) {
```

Comment:
I would rename this as the function talks about _type_ but only takes a string _prefix_.